lib/cretonne/src/ir/instructions.rs

```diff
@@ -499,17 +500,88 @@ impl OpcodeConstraints {
     }
 }
 
+/// A small bitset built on a single primitive integer type
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct BitSet<T>(pub T);
+
+/// Makrer trait for the types that can be used inside a BitSet - u8, u16,
+pub trait BitSetInnerType<T> {
+    /// Needed to coerce 1, 0 to T in fn contains()
+    fn coerce(u8) -> T;
+    /// Needed to limit min/max
+    fn bits() -> usize;
+}
+
+impl BitSetInnerType<u8> for u8 {
+    fn coerce(data: u8) -> u8 { data }
+    fn bits() -> usize { size_of::<u8>()*8 }
+}
+
+impl BitSetInnerType<u16> for u16 {
+    fn coerce(data: u8) -> u16 { data as u16 }
+    fn bits() -> usize { size_of::<u16>()*8 }
+}
+
+impl<T> BitSet<T> where
+    T: BitSetInnerType<T> +
+       BitAnd<T, Output=T> +
+       BitOr<T, Output=T> +
+       Shl<u8, Output=T> +
+       Copy +
+       PartialEq
+{
+    /// Check if this BitSet contains the number num
+    pub fn contains(&self, num: u8) -> bool {
+        assert!((num as usize) < T::bits());
```

Comment:
I think `debug_assert!` will suffice here.

It seems that `bits()` doesn't need to be a trait method. Just use `size_of::<T>()*8`.

I actually think that `<<` will trap in debug builds if the shift amount is too large. That may be god enough.

A different approach that may be simpler to handle: Instead of the `BitSetInnerType` trait, just require `T: Into<u32>` and do all of your arithmetic on `u32`. The `Into<u32>` trait is implemented by `u8`, `u16`, and `u32`.